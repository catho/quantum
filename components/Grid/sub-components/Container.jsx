import PropTypes from 'prop-types';
import styled from 'styled-components';
import { hide, query, calcGutter } from './shared';
import { theme as defaultTheme } from '../../shared';
import { CSSVariables } from '../../GlobalStyle';

const renderBreakpoint = (
  { theme: { gutter, breakpoints }, 'no-gutters': noGutters },
  breakpoint,
) => {
  const calculedGutter = CSSVariables({
    theme: {
      gutter: Number(calcGutter(gutter, noGutters, true)),
    },
  }).gutter[breakpoint];
  const q = query(breakpoints)[breakpoint];

  return q`
    padding: ${calculedGutter};
  `;
};

const renderResponsives = ({
  theme: { breakpoints, gutter },
  'no-gutters': noGutters,
}) =>
  Object.keys(breakpoints).map(breakpoint =>
    renderBreakpoint(
      {
        theme: {
          breakpoints,
          gutter,
        },
        'no-gutters': noGutters,
      },
      breakpoint,
    ),
  );

const Container = styled.div`
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;

  ${renderResponsives}
  ${({ fluid }) => !fluid && 'max-width: 95%;'}
  ${hide}

  & & {
    max-width: 100%;
  }
`;

Container.propTypes = {
  fluid: PropTypes.bool,
  hide: PropTypes.oneOfType([
    PropTypes.oneOf(Object.keys(defaultTheme.breakpoints)),
    PropTypes.arrayOf(PropTypes.string),
  ]),
  theme: PropTypes.shape({
    breakpoints: PropTypes.object,
  }),
  'no-gutters': PropTypes.bool,
};

Container.defaultProps = {
  fluid: false,
  'no-gutters': false,
  theme: defaultTheme,
};

Container.displayName = 'Container';

export default Container;
