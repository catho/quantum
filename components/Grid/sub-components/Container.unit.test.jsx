import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import Container from './Container';

describe('<Container />', () => {
  describe('with hide prop', () => {
    it("should hide it's children according to screen size", () => {
      expect(
        renderer.create(<Container hide="xsmall">Content</Container>).toJSON(),
      ).toMatchSnapshot();
      expect(
        renderer.create(<Container hide="small">Content</Container>).toJSON(),
      ).toMatchSnapshot();
      expect(
        renderer.create(<Container hide="medium">Content</Container>).toJSON(),
      ).toMatchSnapshot();
      expect(
        renderer.create(<Container hide="large">Content</Container>).toJSON(),
      ).toMatchSnapshot();
      expect(
        renderer.create(<Container hide="xlarge">Content</Container>).toJSON(),
      ).toMatchSnapshot();
      expect(
        renderer
          .create(<Container hide={['medium', 'xlarge']}>Content</Container>)
          .toJSON(),
      ).toMatchSnapshot();
    });
  });

  describe('with fluid prop', () => {
    it('should have a fluid width', () => {
      expect(
        renderer.create(<Container fluid>Content</Container>).toJSON(),
      ).toMatchSnapshot();
    });
  });

  describe('with withBreakpoints prop', () => {
    it('should have a fixed width', () => {
      expect(
        renderer
          .create(<Container withBreakpoints>Content</Container>)
          .toJSON(),
      ).toMatchSnapshot();
    });
  });

  describe('with "no-gutters" prop', () => {
    it('should not have gutters', () => {
      expect(
        renderer.create(<Container no-gutters>Content</Container>).toJSON(),
      ).toMatchSnapshot();
    });
  });
});
