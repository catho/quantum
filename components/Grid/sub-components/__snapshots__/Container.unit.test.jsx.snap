// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container /> with "no-gutters" prop should not have gutters 1`] = `
.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  max-width: 95%;
}

.c1 .c0 {
  max-width: 100%;
}

@media (min-width:0px) {
  .c1 {
    padding: 0px;
  }
}

@media (min-width:600px) {
  .c1 {
    padding: 0px;
  }
}

@media (min-width:1024px) {
  .c1 {
    padding: 0px;
  }
}

@media (min-width:1440px) {
  .c1 {
    padding: 0px;
  }
}

@media (min-width:1920px) {
  .c1 {
    padding: 0px;
  }
}

<div
  className="c0 c1"
>
  Content
</div>
`;

exports[`<Container /> with fluid prop should have a fluid width 1`] = `
.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
}

.c1 .c0 {
  max-width: 100%;
}

@media (min-width:0px) {
  .c1 {
    padding: 16px;
  }
}

@media (min-width:600px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1024px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1440px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1920px) {
  .c1 {
    padding: 24px;
  }
}

<div
  className="c0 c1"
>
  Content
</div>
`;

exports[`<Container /> with hide prop should hide it's children according to screen size 1`] = `
.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  max-width: 95%;
}

.c1 .c0 {
  max-width: 100%;
}

@media (min-width:0px) {
  .c1 {
    padding: 16px;
  }
}

@media (min-width:600px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1024px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1440px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1920px) {
  .c1 {
    padding: 24px;
  }
}

@media (max-width:599px) {
  .c1 {
    display: none !important;
  }
}

<div
  className="c0 c1"
>
  Content
</div>
`;

exports[`<Container /> with hide prop should hide it's children according to screen size 2`] = `
.c2 .c0 {
  max-width: 100%;
}

.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  max-width: 95%;
}

.c1 .c0 {
  max-width: 100%;
}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (max-width:599px) {

}

@media (min-width:0px) {
  .c1 {
    padding: 16px;
  }
}

@media (min-width:600px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1024px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1440px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1920px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:600px) and (max-width:1023px) {
  .c1 {
    display: none !important;
  }
}

<div
  className="c0 c1"
>
  Content
</div>
`;

exports[`<Container /> with hide prop should hide it's children according to screen size 3`] = `
.c2 .c0 {
  max-width: 100%;
}

.c3 .c0 {
  max-width: 100%;
}

.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  max-width: 95%;
}

.c1 .c0 {
  max-width: 100%;
}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (max-width:599px) {

}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (min-width:600px) and (max-width:1023px) {

}

@media (min-width:0px) {
  .c1 {
    padding: 16px;
  }
}

@media (min-width:600px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1024px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1440px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1920px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1024px) and (max-width:1439px) {
  .c1 {
    display: none !important;
  }
}

<div
  className="c0 c1"
>
  Content
</div>
`;

exports[`<Container /> with hide prop should hide it's children according to screen size 4`] = `
.c2 .c0 {
  max-width: 100%;
}

.c3 .c0 {
  max-width: 100%;
}

.c4 .c0 {
  max-width: 100%;
}

.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  max-width: 95%;
}

.c1 .c0 {
  max-width: 100%;
}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (max-width:599px) {

}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (min-width:600px) and (max-width:1023px) {

}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (min-width:1024px) and (max-width:1439px) {

}

@media (min-width:0px) {
  .c1 {
    padding: 16px;
  }
}

@media (min-width:600px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1024px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1440px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1920px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1440px) and (max-width:1919px) {
  .c1 {
    display: none !important;
  }
}

<div
  className="c0 c1"
>
  Content
</div>
`;

exports[`<Container /> with hide prop should hide it's children according to screen size 5`] = `
.c2 .c0 {
  max-width: 100%;
}

.c3 .c0 {
  max-width: 100%;
}

.c4 .c0 {
  max-width: 100%;
}

.c5 .c0 {
  max-width: 100%;
}

.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  max-width: 95%;
}

.c1 .c0 {
  max-width: 100%;
}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (max-width:599px) {

}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (min-width:600px) and (max-width:1023px) {

}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (min-width:1024px) and (max-width:1439px) {

}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (min-width:1440px) and (max-width:1919px) {

}

@media (min-width:0px) {
  .c1 {
    padding: 16px;
  }
}

@media (min-width:600px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1024px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1440px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1920px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1920px) {
  .c1 {
    display: none !important;
  }
}

<div
  className="c0 c1"
>
  Content
</div>
`;

exports[`<Container /> with hide prop should hide it's children according to screen size 6`] = `
.c2 .c0 {
  max-width: 100%;
}

.c3 .c0 {
  max-width: 100%;
}

.c4 .c0 {
  max-width: 100%;
}

.c5 .c0 {
  max-width: 100%;
}

.c6 .c0 {
  max-width: 100%;
}

.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  max-width: 95%;
}

.c1 .c0 {
  max-width: 100%;
}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (max-width:599px) {

}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (min-width:600px) and (max-width:1023px) {

}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (min-width:1024px) and (max-width:1439px) {

}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (min-width:1440px) and (max-width:1919px) {

}

@media (min-width:0px) {

}

@media (min-width:600px) {

}

@media (min-width:1024px) {

}

@media (min-width:1440px) {

}

@media (min-width:1920px) {

}

@media (min-width:1920px) {

}

@media (min-width:0px) {
  .c1 {
    padding: 16px;
  }
}

@media (min-width:600px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1024px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1440px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1920px) {
  .c1 {
    padding: 24px;
  }
}

@media (min-width:1024px) and (max-width:1439px) {
  .c1 {
    display: none !important;
  }
}

@media (min-width:1920px) {
  .c1 {
    display: none !important;
  }
}

<div
  className="c0 c1"
>
  Content
</div>
`;

exports[`<Container /> with withBreakpoints prop should have a fixed width 1`] = `
.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  max-width: 95%;
}

.c1 .c0 {
  max-width: 100%;
}

@media (min-width:0px) {
  .c1 {
    width: 100%,;
    padding: 16px;
  }
}

@media (min-width:600px) {
  .c1 {
    width: 100%;
    padding: 24px;
  }
}

@media (min-width:1024px) {
  .c1 {
    width: 1024px;
    padding: 24px;
  }
}

@media (min-width:1440px) {
  .c1 {
    width: 1440px;
    padding: 24px;
  }
}

@media (min-width:1920px) {
  .c1 {
    width: 1920px;
    padding: 24px;
  }
}

<div
  className="c0 c1"
>
  Content
</div>
`;
